src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGoogleServicesDAO.scala

```diff
@@ -324,6 +324,31 @@ class HttpGoogleServicesDAO(
     }
   }
 
+  override def getBucketUsage(bucketName: String, maxResults: Option[Long]): Future[BigInt] = {
+    def sumItemSizes(items: java.util.List[StorageObject]) = {
+      (BigInt(0) /: items) { (sum, o) => sum + o.getSize }
+    }
+
+    def processPage(pageToken: Option[String] = None): Future[BigInt] = {
+      val fetcher = getStorage(getBucketServiceAccountCredential).objects().list(bucketName)
+      if (maxResults.isDefined) fetcher setMaxResults maxResults.get
+      if (pageToken.isDefined) fetcher setPageToken pageToken.get
+
+      retryWithRecoverWhen500orGoogleError(() => {
+        val result = executeGoogleRequest(fetcher)
+        result match {
+          case r if r.getItems == null => BigInt(0)
+          case r if r.getNextPageToken == null => sumItemSizes(r.getItems)
+          case r => sumItemSizes(r.getItems) + Await.result(processPage(Some(result.getNextPageToken)), Duration.Inf)
```

Comment:
Unfortunately, it is not fine here either. The problem is that Await blocks the thread which can lead to resource contention/starvation.

The solution is to chain futures together with map or flatMap. I would try something like
```
def processPage(sum: BigInt, pageToken: Option[String] = None): Future[(BigInt, Option[String])] = {
  val fetcher = getStorage(getBucketServiceAccountCredential).objects().list(bucketName)
  maxResults.forEach { fetcher setMaxResults maxResults.get }
  pageToken.forEach { fetcher setPageToken pageToken.get }

  retryWhen500orGoogleError(() => {
    val result = executeGoogleRequest(fetcher)
    (Option(result.getItems).map(sumItemSizes(_.getItems)).getOrElse(BigInt(0)), Option(result.getNextPageToken))
  }) flatMap { 
    case (pageSize, None) => Future.successful(sum + pageSize)
    case (pageSize, Some(nextPageToken)) => processPage(sum + pageSize, Option(nextPageToken))
  }
}
```